### PR TITLE
use blueprint specified template folder

### PR DESCRIPTION
**Pull request by alanhamlett**
Fixes issue #1361.

### CODE REVIEW

File: flask/templating.py [hunk 1]
```diff
@@ -121,8 +121,24 @@ def render_template(template_name_or_list, **context):
...
     ctx = _app_ctx_stack.top
     ctx.app.update_template_context(context)
-    return _render(ctx.app.jinja_env.get_or_select_template(template_name_or_list),
-                   context, ctx.app)
+
+    template = None
+
+    if request.blueprint is not None and \
+            isinstance(template_name_or_list, string_types):
+        bp = current_app.blueprints[request.blueprint]
```
Comment (untitaker) [inline flask/templating.py]: Use `ctx.app`, it's already in scope.

Comment (untitaker) [inline flask/templating.py]: Also `_request_ctx_stack` contains the request, therefore no need to import `current_app` and `request`.

File: flask/templating.py [hunk 2]
```diff
@@ -112,7 +113,6 @@ def _render(template, context, app):
 def render_template(template_name_or_list, **context):
     """Renders a template from the template folder with the given
     context.
-
```
Comment (untitaker) [inline flask/templating.py]: Why remove this line?

Comment (alanhamlett) [inline flask/templating.py]: My editor automatically trims whitespace from end of lines.

Comment (untitaker) [inline flask/templating.py]: Please undo this, I think this actually breaks the rst.

**Comment by svieira** [comment]
Unless I am greatly mistaken this will break anyone out there who is using extensions and overriding (not inheriting from) the extension's blueprints.  For example, given the following extension setup:

```text
extension/
    __init__.py
    some.py
    extension_details.py
    templates/
        extension/
            base.html
            some-page.html
```

before this patch, overriding `base.html` would simply have been a matter of:

```text
the-app/
    __init__.py
    templates/
        the-layout.html
        extension/
            base.html
```

And making sure that `base.html@the-app` had the necessary `block` and `import` declarations.  This would shadow `base.html@extension`, allowing the app developer to override what the extension was doing without the extension developer having to provide a hook for such an overload

With this patch, I don't think there is a way to override `base.html@extension` in `extension` without patching Flask to bring back the old behavior (which will result in a lot of issues if new extensions start depending on this behavior).

**Comment by davidism** [comment]
Agree with @svieira, this is the point I was making in [my comment on the original issue](https://github.com/mitsuhiko/flask/issues/1361#issuecomment-121746515).

> The point of searching the app template folder first is so that an app developer can override templates supplied by a blueprint from an extension.

**Comment by ThiefMaster** [comment]
Those commits should be squashed into a single one eventually.

Also, this would be a breaking change depending on the application structure, so I'd rather make it configurable, with the current behavior being the default.

**Comment by svieira** [comment]
If configurable it would have to be done at the `register_blueprint` method (since you could have one extension looking for the old behavior and one needed the new behavior).

I created #1548 to talk about some another possibility that might solve several of these issues in one go.

**Comment by AzureusNation** [comment]
Is there any update on when this fix could be merged or implemented in a different manner?

**Comment by davidism** [comment]
The referenced issue was closed as wontfix.

**Pull request closed without merging**